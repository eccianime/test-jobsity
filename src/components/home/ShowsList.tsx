import { SHOWS_DATA } from "@/data/show_list";
import { memo } from "react";
import { FlatList, View } from "react-native";
import ShowsListFooter from "./ShowsListFooter";
import ShowItem from "./ShowsListItem";

function ShowsList() {
  return (
    <View className="flex-1 p-4">
      <FlatList
        showsVerticalScrollIndicator={false}
        numColumns={2}
        keyExtractor={(item) => item.id.toString()}
        contentContainerClassName="gap-4"
        data={SHOWS_DATA}
        removeClippedSubviews
        renderItem={({ item, index }) => <ShowItem data={item} index={index} />}
        ListFooterComponent={
          <ShowsListFooter currentPage={1} handleChangePage={() => {}} />
        }
      />
    </View>
  );
}

export default memo(ShowsList);
