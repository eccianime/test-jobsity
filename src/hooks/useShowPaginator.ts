import { useLazyGetShowsQuery } from "@/redux/services/shows";
import { ShowProps } from "@/types/schema";
import { useCallback, useEffect, useState } from "react";

export default function useShowPaginator() {
  const [apiPage, setApiPage] = useState(0);
  const [uiPage, setUiPage] = useState(1);
  const [dataCache, setDataCache] = useState<ShowProps[]>([]);
  const [hasMore, setHasMore] = useState(true);
  const [isLoading, setIsLoading] = useState(false);

  const [getShows] = useLazyGetShowsQuery();

  const loadMoreFromApi = useCallback(async () => {
    if (!hasMore || isLoading) return;

    setIsLoading(true);

    const response = await getShows({ page: apiPage });

    setDataCache((prev) => [...prev, ...(response.data as ShowProps[])]);

    const responseToHasMore = await getShows({ page: apiPage + 1 });

    if (responseToHasMore?.data?.length === 0) {
      setHasMore(false);
    }

    setApiPage((prev) => prev + 1);
    setIsLoading(false);
  }, [hasMore, isLoading, apiPage, getShows]);

  const getCurrentPageData = useCallback((page: number) => {
    setUiPage(page);
  }, []);

  useEffect(() => {
    const start = (uiPage - 1) * 10;
    const end = start + 10;

    if (end > dataCache.length && hasMore && !isLoading) {
      loadMoreFromApi();
    }
  }, [uiPage, dataCache.length, hasMore, isLoading, loadMoreFromApi]);

  const getPaginationNumbers = useCallback(() => {
    if (uiPage === 1) {
      return [1, 2, 3];
    }
    const maxUiPage = Math.ceil(dataCache.length / 10);
    if (!hasMore && uiPage >= maxUiPage) {
      return [maxUiPage - 2, maxUiPage - 1, maxUiPage].filter((p) => p > 0);
    }
    return [uiPage - 1, uiPage, uiPage + 1];
  }, [uiPage, dataCache.length, hasMore]);

  const currentData = dataCache.slice((uiPage - 1) * 10, uiPage * 10);

  return {
    currentData,
    uiPage,
    goToPage: getCurrentPageData,
    paginationNumbers: getPaginationNumbers(),
    hasMore,
    isLoading,
  };
}
