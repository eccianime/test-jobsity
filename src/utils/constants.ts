export const COMPONENTS_TEST_IDS = {
  BACK_BUTTON: "SRC_COMPONENTS_BACKBUTTON",
};
