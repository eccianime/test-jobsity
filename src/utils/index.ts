export function removeHtmlTags(htmlString: string): string {
  const regex = /<.*?>/g;
  return htmlString.replace(regex, "");
}
