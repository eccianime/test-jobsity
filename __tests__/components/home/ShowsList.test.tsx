import ShowsList from "@/components/home/ShowsList";
import useShowPaginator from "@/hooks/useShowPaginator";
import { COMPONENTS_TEST_IDS } from "@/utils/constants";
import { render } from "@testing-library/react-native";

jest.mock("@/hooks/useShowPaginator");

jest.mock("@/components/LoadingScreen", () => {
  const MockLoadingScreen = () => <></>;
  MockLoadingScreen.displayName = "LoadingScreen";
  return MockLoadingScreen;
});

jest.mock("@/components/home/ShowsListFooter", () => {
  const MockShowsListFooter = () => <></>;
  MockShowsListFooter.displayName = "ShowsListFooter";
  return MockShowsListFooter;
});

jest.mock("@/components/home/ShowsListItem", () => {
  const MockShowsListItem = () => <></>;
  MockShowsListItem.displayName = "ShowsListItem";
  return MockShowsListItem;
});

describe("ShowsList", () => {
  it("should render LoadingScreen when isLoading is true", () => {
    (useShowPaginator as jest.Mock).mockReturnValue({
      isLoading: true,
      currentData: [],
      paginationNumbers: [],
      uiPage: 1,
      hasMore: true,
      goToPage: jest.fn(),
    });
    const { toJSON } = render(<ShowsList />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("should render FlatList with correct props", () => {
    const mockData = [
      { id: 1, name: "Show 1", image: {} },
      { id: 2, name: "Show 2", image: {} },
    ];
    (useShowPaginator as jest.Mock).mockReturnValue({
      isLoading: false,
      currentData: mockData,
      paginationNumbers: [1, 2],
      uiPage: 1,
      hasMore: true,
      goToPage: jest.fn(),
    });

    const { getByTestId } = render(<ShowsList />);
    expect(getByTestId(COMPONENTS_TEST_IDS.SHOWS_LIST)).toBeTruthy();
  });
});
