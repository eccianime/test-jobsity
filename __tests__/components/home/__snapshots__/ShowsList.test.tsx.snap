// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowsList should render LoadingScreen when isLoading is true 1`] = `null`;
